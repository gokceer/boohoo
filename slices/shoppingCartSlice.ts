import { createSlice } from "@reduxjs/toolkit";

import type { RootState } from '../store'
import type { CommonProductObjectType } from "../App"

type initialStateType = {
  items: CommonProductObjectType[]
}

const initialState: initialStateType = {
  items: [],
};

export const shoppingCartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    addToCart: (state, action) => {
      state.items = [...state.items, action.payload];
    },
    removeFromCart: (state, action) => {
      const index = state.items.findIndex((item) => item.id === action.payload.id);
        
      let tempCart = [...state.items];
        
      index >= 0
        ? tempCart.splice(index, 1)
        : console.warn(`Cannot remove id: ${action.payload.id}`);
        
      state.items = tempCart;
    },
    removeAllFromCartWithId: (state, action) => {
      const filteredState = state.items.filter((item) => {
        return item.id !== action.payload.id
      })
      state.items = [...filteredState]
    },
  }
});
  
  export const { addToCart, removeFromCart, removeAllFromCartWithId } = shoppingCartSlice.actions;
  
  export const selectCartItems = (state: RootState) => state.cart.items;
  
  export const selectCartItemsWithId = (state: RootState, id: number) =>
  state.cart.items.filter((item) => item.id === id);
  
  export const selectCartTotal = (state: RootState) =>
  state.cart.items.reduce((total, item) => (total += item.price), 0);
  
  export default shoppingCartSlice.reducer;
