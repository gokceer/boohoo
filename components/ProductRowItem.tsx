import React from 'react'
import { View, Text, StyleSheet, TouchableOpacity, Image } from 'react-native'
import { useAppDispatch } from '../hooks'
import { removeAllFromCartWithId } from "../slices/shoppingCartSlice"
import EditQuantity from '../components/EditQuantity';

import type { CommonProductObjectType } from "../App"

type ProductRowItemPropsType = {
  items: CommonProductObjectType[]
}

const ProductRowItem = ({ items }: ProductRowItemPropsType) => {
  const dispatch = useAppDispatch()

  // Display product information, quantity and price in a row. Edit quantity/remove product.
  return (
    <View style={styles.container}>
      <View style={{ alignItems: "center" }}>
        <EditQuantity itemInfo={items[0]} />
        <TouchableOpacity>
          <Text
            style={{ color: "#841584" }}
            onPress={() => dispatch(removeAllFromCartWithId({ id: items[0]?.id }))}
          >
            Remove
          </Text>
        </TouchableOpacity>
      </View>

      <Image style={styles.image} source={{ uri: items[0]?.img }} />
      <Text style={{ flex: 1 }} numberOfLines={4} ellipsizeMode='tail'>{items[0]?.name}</Text>
      <Text>
        Â£ {items[0]?.price * items.length}
      </Text>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    paddingVertical: 10,
    paddingHorizontal: 5
  },
  image: {
    flex: 1,
    width: '100%',
    aspectRatio: 1,
    resizeMode: 'contain'
  },
})

export default ProductRowItem