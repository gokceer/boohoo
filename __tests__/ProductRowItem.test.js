import { Provider } from "react-redux";
import { store } from "../store";
import { render, fireEvent, act } from "@testing-library/react-native";
import ProductRowItem from "../components/ProductRowItem";
import { addToCart } from "../slices/shoppingCartSlice";

const mockItemProp = {
  id: 1,
  colour: "Black",
  name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
  price: 10,
  img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
};

const mockItemPropArr = [mockItemProp];

jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");

const setup = (props = {}) => {
  const wrapper = render(
    <Provider store={store}>
      <ProductRowItem {...props} />
    </Provider>
  );
  return wrapper;
};

describe("Product Row Item unit tests", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it("renders correctly, snapshot test", () => {
    const tree = setup({ items: mockItemPropArr });
    expect(tree).toMatchSnapshot();
  });

  it("All parts in the row existance test", () => {
    const wrapper = setup({ items: mockItemPropArr });

    expect(wrapper.getByText("Remove")).toBeTruthy();
    expect(wrapper.getByTestId("product_img")).toBeTruthy();
    expect(
      wrapper.getByText("Black Sheet Strappy Textured Glitter Bodycon Dress")
    ).toBeTruthy();
    expect(wrapper.getByText("Â£ 10")).toBeTruthy();
  });

  it("Remove items by pressing Remove test", () => {
    // add item to the store
    act(() => {
      store.dispatch(addToCart(mockItemProp));
    });

    const wrapper = setup({ items: mockItemPropArr });
    const removeButton = wrapper.getByText("Remove");

    fireEvent.press(removeButton);

    expect(store.getState().cart.items).toEqual([]);
  });
});
