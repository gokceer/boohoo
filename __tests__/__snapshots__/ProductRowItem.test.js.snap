// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Row Item unit tests renders correctly, snapshot test 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "paddingHorizontal": 5,
      "paddingVertical": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="decrement_btn"
        >
          <RNSVGSvgView
            align="xMidYMid"
            aria-hidden="true"
            bbHeight="32"
            bbWidth="32"
            color="gray"
            fill="currentColor"
            focusable={false}
            height={32}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 32,
                  "width": 32,
                },
              ]
            }
            tintColor="gray"
            vbHeight={24}
            vbWidth={24}
            width={32}
          >
            <RNSVGGroup
              fill={
                Object {
                  "type": 2,
                }
              }
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGPath
                clipRule={0}
                d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25zm3 10.5a.75.75 0 000-1.5H9a.75.75 0 000 1.5h6z"
                fillRule={0}
                propList={
                  Array [
                    "fillRule",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
        <Text>
          0
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="increment_btn"
        >
          <RNSVGSvgView
            align="xMidYMid"
            aria-hidden="true"
            bbHeight="32"
            bbWidth="32"
            color="gray"
            fill="currentColor"
            focusable={false}
            height={32}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 32,
                  "width": 32,
                },
              ]
            }
            tintColor="gray"
            vbHeight={24}
            vbWidth={24}
            width={32}
          >
            <RNSVGGroup
              fill={
                Object {
                  "type": 2,
                }
              }
              propList={
                Array [
                  "fill",
                ]
              }
            >
              <RNSVGPath
                clipRule={0}
                d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25zM12.75 9a.75.75 0 00-1.5 0v2.25H9a.75.75 0 000 1.5h2.25V15a.75.75 0 001.5 0v-2.25H15a.75.75 0 000-1.5h-2.25V9z"
                fillRule={0}
                propList={
                  Array [
                    "fillRule",
                  ]
                }
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        onPress={[Function]}
        style={
          Object {
            "color": "#841584",
          }
        }
      >
        Remove
      </Text>
    </View>
  </View>
  <Image
    source={
      Object {
        "uri": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
      }
    }
    style={
      Object {
        "aspectRatio": 1,
        "flex": 1,
        "resizeMode": "contain",
        "width": "100%",
      }
    }
    testID="product_img"
  />
  <Text
    ellipsizeMode="tail"
    numberOfLines={4}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    Black Sheet Strappy Textured Glitter Bodycon Dress
  </Text>
  <Text>
    Â£ 
    10
  </Text>
</View>
`;
